### PR TITLE
Empty listing Bug Fix and Other Pending Changes

### PR DESCRIPTION
**Pull request by anujmodi2021**

### CODE REVIEW

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java:262]: Why are we removing the close() method call here ?

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/AbstractAbfsIntegrationTest.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/AbstractAbfsIntegrationTest.java]: double semicolon

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/UriUtils.java
Comment (anmolanmol1234) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/UriUtils.java]: Is it possible to have more than one .blob. in the url ?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java:262]: We needed close because we were checking it after creating filesystem.
Now we are checking it before creating file system and failing, so no need to close.

File: hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/AbstractAbfsIntegrationTest.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/AbstractAbfsIntegrationTest.java]: Taken

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/UriUtils.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/UriUtils.java]: Updated code as per the discussion offline

**Review by anmolanmol1234 — APPROVED** [review]

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java:262]: Since this is a significant change will trigger a CI.
We can wait for CI results before merging this.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/UriUtils.java
Comment (anujmodi2021) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/utils/UriUtils.java]: These two can be moved inside replacedUrl(0 function only

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java
Comment (snvijaya) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java:649]: if (continuation != null) {
      abfsUriQueryBuilder.addQuery(QUERY_PARAM_MARKER, continuation);
    }

**Review by snvijaya — APPROVED** [review]
One minor comment in AbfsBlobClient. Please take a look before checkin.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java
Comment (anujmodi2021) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsBlobClient.java:649]: This was a redundant check. Same check is already added in AbfsUriQueryBuilder.java
```
public void addQuery(final String name, final String value) {
    if (value != null && !value.isEmpty()) {
      this.parameters.put(name, value);
    }
  }
```

**Comment by anujmodi2021** [comment]
------------------------------
:::: AGGREGATED TEST RESULT ::::

============================================================
HNS-OAuth-DFS
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 4
[WARNING] Tests run: 784, Failures: 0, Errors: 0, Skipped: 179
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 51

============================================================
HNS-SharedKey-DFS
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 5
[WARNING] Tests run: 784, Failures: 0, Errors: 0, Skipped: 131
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 38

============================================================
NonHNS-SharedKey-DFS
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 768, Failures: 0, Errors: 0, Skipped: 371
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 39

============================================================
AppendBlob-HNS-OAuth-DFS
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 4
[WARNING] Tests run: 784, Failures: 0, Errors: 0, Skipped: 182
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 77

============================================================
NonHNS-SharedKey-Blob
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 11
[WARNING] Tests run: 768, Failures: 0, Errors: 0, Skipped: 285
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 39

============================================================
NonHNS-OAuth-DFS
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 768, Failures: 0, Errors: 0, Skipped: 376
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 52

============================================================
NonHNS-OAuth-Blob
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 11
[ERROR] Tests run: 768, Failures: 0, Errors: 0, Skipped: 289
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 52

============================================================
AppendBlob-NonHNS-OAuth-Blob
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 11
[WARNING] Tests run: 768, Failures: 0, Errors: 0, Skipped: 306
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 78

============================================================
HNS-Oauth-DFS-IngressBlob
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 4
[WARNING] Tests run: 784, Failures: 0, Errors: 0, Skipped: 308
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 51

============================================================
NonHNS-OAuth-DFS-IngressBlob
============================================================

[WARNING] Tests run: 163, Failures: 0, Errors: 0, Skipped: 10
[WARNING] Tests run: 768, Failures: 0, Errors: 0, Skipped: 372
[WARNING] Tests run: 440, Failures: 0, Errors: 0, Skipped: 52

Time taken: 183 mins 4 secs.

**Pull request merged**
